### ISSUE TITLE
dom::object::iterator missing some expected operators.

### ISSUE THREAD

**Issue by TkTech** [simdjson/simdjson#1046]
`dom::object::iterator` is missing a few operators that make it incompatible with anything expecting std::iterator.

In particular:

- `operator!=` is present, but not `operator==` (breaking pybind11's `make_iterator`)
- `iterator operator++(int)`

**Comment by lemire** [comment]
Marked for 0.5.

**Comment by TkTech** [comment]
Same with `dom::array::iterator`, temporarily patched downstream in pysimdjson's simdjson.h.

**Comment by TkTech** [comment]
It's also missing a copy constructor, so wrapping the iterator is really annoying.

It's pretty common for wrappers around objects to have the equivalent of Python's keys(), values(), and items(). items() is what the default dom::object::iterator implements (and that's sensible). keys() gives you just the keys, values() just the values.

But say if I want to make a new KeyIterator, that only returns the keys, I have to re-implement everything instead of just being able to do something like (psuedo-code):

```
class KeyIterator : public dom::object::iterator {
    public:
        KeyIterator(dom::object::iterator it) : dom::object::iterator(it) {};

        PyObject operator*() {
            return sv_to_unicode(dom::object::iterator::operator*().key);
        }
};
```

They're just missing a lot of the things a template library will expect an iterator to have. Making `key_value_pair` a real std::pair would also make life easier, since it's what STL algorithms will expect.

**Comment by TkTech** [comment]
Closed by #1050 thanks to @llllllllll.
